repository-data/application/src/main/resources/hcm-config/configuration/update/Migration20171208Resource.groovy

```diff
@@ -0,0 +1,61 @@
+package org.hippoecm.frontend.plugins.cms.admin.updater
+
+import javax.jcr.*
+import org.onehippo.repository.update.BaseNodeUpdateVisitor
+
+
+class Migration20171208Resource extends BaseNodeUpdateVisitor {
+
+    boolean doUpdate(Node node) {
+        try {
+            log.warn("Node type: " + node.getPrimaryNodeType().getName());
+            if ("publicationsystem:publication".equals(node.getPrimaryNodeType().getName())) {
+                return renameSubnode(node, "publicationsystem:attachments", "publicationsystem:resource",
+                    "publicationsystem:Attachments-v2");
+            } else if ("publicationsystem:dataset".equals(node.getPrimaryNodeType().getName())) {
+                return renameSubnode(node, "publicationsystem:Files", "publicationsystem:resource",
+                    "publicationsystem:Files-v2");
+            }
+
+        } catch (e) {
+            log.error("Failed to process record.", e)
+        }
+
+        return false
+    }
+
+    boolean undoUpdate(Node node) {
+        try {
+            log.warn("Node type: " + node.getPrimaryNodeType().getName());
+            if ("publicationsystem:publication".equals(node.getPrimaryNodeType().getName())) {
+                return renameSubnode(node, "publicationsystem:Attachments-v2", "hippo:resource",
+                    "publicationsystem:attachments");
+            } else if ("publicationsystem:dataset".equals(node.getPrimaryNodeType().getName())) {
+                return renameSubnode(node, "publicationsystem:Files-v2", "hippo:resource",
+                    "publicationsystem:Files");
+            }
+
+        } catch (e) {
+            log.error("Failed to process record.", e)
+        }
+
+        return false
+    }
+
+    boolean renameSubnode(Node node, String from, String toType, String toName) {
+        NodeIterator i = node.getNodes()
+        Node n;
+
+        while(i.hasNext()) {
```

Comment:
Consider rewriting as `for (NodeIterator i = node.getNodes(); i.hasNext(); ) { n = i.nextNode(); ...}`?